api/envoy/config/cluster/v3/cluster.proto

```diff
@@ -499,6 +499,15 @@ message Cluster {
     google.protobuf.Duration max_interval = 2 [(validate.rules).duration = {gt {nanos: 1000000}}];
   }
 
+  // Allows creation and configuration of custom upstream type.
+  message UpstreamType {
+    // Required. The name of the upstream type
+    string name = 1;
```

Comment:
You can use `TypedExtensionConfig` for this message and probably just remove `UpstreamType`